### ISSUE TITLE
[RFC] Fine-grained control over runtime

### ISSUE THREAD

**Issue by pavlukivan** [RustPython/RustPython#2370]
## Summary

For running untrusted code, precise control over what system access is available to the user is needed. This mostly means controlling file io (and, by extension, module loading).

## Detailed Explanation

The way I see this usually done is allowing the user to register custom module loaders to let them control the entire module loading process. They can decide whether to allow access to system modules, lazily load their own modules, load (or not load) user modules from the filesystem or from RAM, and so on. Some scripting engines provide control over io or memory allocation, or the ability to set interrupt handlers to run at certain periods of time and possibly terminate the program (which is important in singlethreaded environments), but allowing the user to replace or remove system modules is enough for most use cases.

## Drawbacks, Rationale, and Alternatives

The main drawback is an ABI change, there's a lot of restructuring required from what I can tell. This will help a lot with platform-specific support (e.g. wasm) by separating the module logic from filesystem access, as well as allow developers to use the project in untrusted environments without using WASI (that can potentially add a lot of boilerplate and some overhead

## Unresolved Questions

The API: should the developers have full control over file input or just module loading (I think the latter is enough if there's a module override API); how to allow changing the builtins (`open`); how to make small modifications to the stdlib (i.e. make the files load from RAM instead of the filesystem) easy to do.

**Comment by carbotaniuman** [comment]
Iirc there's a Python API/PEP for this, but I don't know it off the top of my head. (Unless I'm misremembering the already implemented import hooks PEP?)

**Comment by pavlukivan** [comment]
You're right, it seems to be part of [importlib](https://docs.python.org/3/library/importlib.html).
